### ISSUE TITLE
'low_section_speed', 'high_section_speed','molten_volume','mold_code', 'cast_pressure','biscuit_thickness','sleeve_temperature'

### ISSUE THREAD

**Issue by ppuppou** [seongjun25/project_diecasting#2]
import pandas as pd

train = pd.read_csv('./train.csv')
train.info()
train.head()
train.columns

df = train[['low_section_speed', 'high_section_speed','molten_volume','mold_code',
            'cast_pressure','biscuit_thickness','sleeve_temperature','passorfail']]
df.loc[df.low_section_speed.isna(),:]
df.info()

import seaborn as sns
import matplotlib.pyplot as plt
plt.rcParams['font.family'] = 'Malgun Gothic'
plt.rcParams['axes.unicode_minus'] = False
feature_columns = df.select_dtypes(include='number').columns.drop('passorfail')
for col in feature_columns:
    fig, axes = plt.subplots(1, 2, figsize=(15, 6))
    fig.suptitle(f"'{col}' 변수의 Pass/Fail 그룹 비교", fontsize=16)
    sns.kdeplot(data=df, x=col, hue='passorfail', 
                fill=True, common_norm=False, palette='Set1', ax=axes[0])
    axes[0].set_title('데이터 분포 비교 (KDE Plot)')
    axes[0].set_xlabel(f'{col} 값')
    axes[0].set_ylabel('밀도 (Density)')
    axes[0].grid(True, linestyle='--', alpha=0.6)
    sns.boxplot(data=df, x='passorfail', y=col, palette='Set1', ax=axes[1])
    axes[1].set_title('데이터 요약 비교 (Box Plot)')
    axes[1].set_xlabel('Pass/Fail')
    axes[1].set_ylabel(f'{col} 값')
    axes[1].grid(True, linestyle='--', alpha=0.6)
    plt.tight_layout(rect=[0, 0.03, 1, 0.95])
    plt.show()

target_column = 'sleeve_temperature'
# 1. passorfail = 0 그룹의 이상치 추출
Q1_0 = df.loc[df['passorfail'] == 0, target_column].quantile(0.25)
Q3_0 = df.loc[df['passorfail'] == 0, target_column].quantile(0.75)
IQR_0 = Q3_0 - Q1_0
lower_bound_0 = Q1_0 - 1.5 * IQR_0
upper_bound_0 = Q3_0 + 1.5 * IQR_0
outliers_pass0 = df.loc[
    (df['passorfail'] == 0) & 
    ((df[target_column] < lower_bound_0) | (df[target_column] > upper_bound_0)),
    target_column
].dropna()
# 2. passorfail = 1 그룹의 이상치 추출
Q1_1 = df.loc[df['passorfail'] == 1, target_column].quantile(0.25)
Q3_1 = df.loc[df['passorfail'] == 1, target_column].quantile(0.75)
IQR_1 = Q3_1 - Q1_1
lower_bound_1 = Q1_1 - 1.5 * IQR_1
upper_bound_1 = Q3_1 + 1.5 * IQR_1
outliers_pass1 = df.loc[
    (df['passorfail'] == 1) & 
    ((df[target_column] < lower_bound_1) | (df[target_column] > upper_bound_1)),
    target_column
].dropna()
# 3. 이상치 데이터 통합
outlier_df_combined = pd.DataFrame({
    target_column: pd.concat([outliers_pass0, outliers_pass1]),
    'passorfail': [0.0] * len(outliers_pass0) + [1.0] * len(outliers_pass1)
})
# 4. 시각화 (KDE Plot만 사용)
plt.figure(figsize=(10, 6))
if not outlier_df_combined.empty:
    sns.kdeplot(data=outlier_df_combined, x=target_column, hue='passorfail', 
                fill=True, common_norm=False, palette='Set1')
    plt.title(f"'{target_column}' 변수의 Pass/Fail 별 '이상치' 분포", fontsize=16)
else:
    plt.title(f"'{target_column}' 변수에는 이상치가 없습니다.", fontsize=16)
plt.xlabel(f'{target_column} 값', fontsize=12)
plt.ylabel('밀도 (Density)', fontsize=12)
plt.grid(True, linestyle='--', alpha=0.6)
plt.legend(title='Pass/Fail')
plt.show()

corr_matrix = df.corr()
plt.figure(figsize=(10, 8))
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', fmt='.2f')
plt.title('변수 간 상관관계 히트맵', fontsize=16)
plt.show()

from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import KNNImputer
model_df = df.copy()
cols_to_check = model_df.columns.drop('molten_volume')
model_df = model_df.dropna(subset=cols_to_check)
imputer = KNNImputer(n_neighbors=5).set_output(transform='pandas')
model = imputer.fit_transform(model_df)
# 독립 변수(X)와 종속 변수(y) 분리
X = model_df.drop('passorfail', axis=1)
y = model_df['passorfail']

# 2. 랜덤 포레스트 모델 학습
# n_jobs=-1: 컴퓨터의 모든 CPU 코어를 사용하여 학습 속도를 높입니다.
# random_state=42: 실행할 때마다 동일한 결과를 얻기 위해 설정합니다.
model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X, y)
# 3. 변수 중요도 추출 및 시각화
# 변수 중요도를 pandas Series로 변환합니다.
importances = pd.Series(model.feature_importances_, index=X.columns)
# 중요도가 높은 순으로 정렬합니다.
importances_sorted = importances.sort_values(ascending=False)
# 시각화
plt.figure(figsize=(12, 8))
sns.barplot(x=importances_sorted, y=importances_sorted.index)
plt.title('랜덤 포레스트 모델의 변수 중요도 (Feature Importance)', fontsize=16)
plt.xlabel('중요도 (Importance)', fontsize=12)
plt.ylabel('변수 (Features)', fontsize=12)
plt.show()
